### DISCUSSION TITLE
Feature: Quota for auto-downloading content and offline syncing playlists

### DISCUSSION THREAD

**Discussion by sandreas** [advplyr/audiobookshelf-app#846] [Ideas]
Hey @advplyr,

recently I thought about offline usage / downloading content. 

So, normally, I set up my app using my personal domain, lets say http://audiobookshelf.example.com and it is always available so no need to care about offline usage. 

But let's say I'm listening to a book and suddenly I have no access to the server (long tunnel when travelling by train or the server crashed, or whatever). To overcome this, I have to download the audio to local media before I know this might happen. Now I have a few other problems:

- If there is no wifi, it'll eat up my available bandwidth
- I have to click "download" for every audio book I would like to have available offline
- If there is no space left on the device, I have to manually delete another book and then go back to download

So here is my idea:

## How about introducing Auto-Download Feature with Quota:

- [ ] Add a global setting: Enable auto download for `Continue Listening` books
- [ ] Add a global setting: Define auto download quota for auto download, default 10GB
- [ ] Add a global setting: Delete finished books after X days, default 30 days
- [ ] Automatically download every audio book, that has been started, has progress, but is not finished - with Option: When Wifi is available
- [ ] If you start listening a new book, it checks the quota and auto-manages the content
   - if there is enough free space, start downloading in background immediately without any warning or visible message
   - if there is NOT enough free space, either there are finished books older than 30 days, so delete them and behave like there is enough free space OR abs shows a dialog `Quota exceeded`where you could
     - Option 1: Button `Extend quota` by the size of the started audio book and behave like there is enough free space
     - Option 2: List auto downloaded files, ordered by `finishedDate ASC, startedDate ASC` with a `remove` button for every item, a `Just stream`button and a cancel button

## Extension Feature: Offline playlists and Favorites playlist

Another Problem I have is, that I have some audio books I'd like to listen now and then, because I just love them. I listened to them many many times and going to do so many times more in the future. I have a Playlist `Favorites` exactly for this purpose. The problem: I still have to download all the content manually. Not that hard, but here is the thing: From time to time, there is another book I would like to listen to again - I add it to the playlist but forget to download it. So managing the content that I want to be available offline is harder than it could be.

- [ ] Add a checkbox to Playlists `sync offline`
- [ ] Automatically download the playlists with the `sync offline` checked
- [ ] Add a "virtual" playlist: `My favorites` that cannot be deleted, but edited to have `sync offline`
- [ ] Add a "Heart" button to all listings of audio books
- [ ] Clicking the heart is a shortcut for adding it to `My favorites`

This would drastically improve usability for many of my use cases.

What do you think? Of course these features could be implemented different or partly, but I think the main point is not having to deal with managing your offline content so much.

**Comment by advplyr** [comment]
Hey, for the auto-download quota I'm pretty sure there are feature requests related to this. I agree we should at least have an option to queue downloads and choose to only download on wifi. Currently there are some issues with running background tasks that are a big hang up right now.

The offline support for Abs in general needs a lot of work. When you are offline you won't see anything but the home page. Storing playlists, collections and series data locally in addition to on the server is a big project. Also, for the library page we have to write all of the sorting/filtering locally to mirror the sorting/filtering that happens on the server

**Comment by sandreas** [comment]
> Hey, for the auto-download quota I'm pretty sure there are feature requests related to this.

Yeah I found it, sorry - and I added my thoughts there again. See #774 

> The offline support for Abs in general needs a lot of work. When you are offline you won't see anything but the home page.

I agree, there is a lot of work to do. Although I think that ABS is a complicated beast - even the API is monstrous. You did a lot of awesome work here but maybe there is room for simplification.

> Storing playlists, collections and series data locally in addition to on the server is a big project. 

I wouldn't... I would start with a simple checkbox in playlists, the API and just the automatic download. That's it. The advantage is, that this is simple and other clients could implement the sync without any further hacks. (I would do so in `ToneAudioPlayer`)

> Also, for the library page we have to write all of the sorting/filtering locally to mirror the sorting/filtering that happens on the server

I don't think that this should be too much work. Did you consider my `DataSource` approach from https://github.com/advplyr/audiobookshelf-app/discussions/428#discussioncomment-6839583? That could include a `LocalDataSource` with the same interface like the others. There could be even a `CachedRemoteDataSource`, that it chaining the `LocalDataSource` within the `AudiobookshelfDataSource`, so that it first looks in the local cache and if it is not there, it downloads the file and begins to stream it.

Whatever, I think you are doing a great job so far and hope these features get implemented some day ;) Any feedback on this?
